Improve error message for match on structs

ergl opened the pull request:
Since a struct doesn't have a type descriptor, it's impossible to match
against it, unless it's exactly the same struct type. Before, the
return type of matchtype.c:is_matchtype didn't convey enough information
to the caller about the reason why two types couldn't match. In most
cases, this is due to reference capabilities, so that's what the error
message shows.

By distinguishing the specific case of a match failure due to lack of
type descriptor, the compiler can give a friendlier error message to the
user.

---

This fixes #2000 and fixes #3565, and mostly follows the approach described by @jemc on the second issue.

jemc reviewed (approved):
Nice work!

ergl commented on src/libponyc/type/matchtype.c:
```diff
@@ -858,13 +907,16 @@ static matchtype_t is_x_match_x(ast_t* operand, ast_t* pattern,
-      return MATCHTYPE_DENY;
+      return MATCHTYPE_DENY_CAP;
+
+    case TK_STRUCT:
+      return MATCHTYPE_DENY_NODESC;
```
I'm not completely sure that this is needed.

ergl replied:
Double checked that this is not needed, so I've removed it.

SeanTAllen commented:
@ergl should this be merged?

also, do you think this deserves a changelog entry and release notes?

ergl commented:
@SeanTAllen Yes, I think it's ready to be merged. As for release notes, sound good, this is probably worth mentioning.

ponylang-main commented:
<!-- release-note-reminder-bot -->
Hi @ergl,

The changelog - changed label was added to this pull request; all PRs with a changelog label need to have release notes included as part of the PR. If you haven't added release notes already, please do.

Release notes are added by creating a uniquely named file in the `.release-notes` directory. We suggest you call the file `3746.md` to match the number of this pull request.

The basic format of the release notes (using markdown) should be:

```
## Title

End user description of changes, why it's important,
problems it solves etc.

If a breaking change, make sure to include 1 or more
examples what code would look like prior to this change
and how to update it to work after this change.
```

Thanks.

SeanTAllen commented:
This commit can be merged once @ergl adds release notes. I'm adding a "do not merge" label until then.

SeanTAllen commented:
@ergl this has a merge conflict that will need to be resolved in addition to the addition of release notes.

ergl commented:
Rebased and added release notes. I've added an example of the new error message in the release notes, but perhaps that's too much? I'm willing to trim it back to just the first paragraph.

SeanTAllen commented:
@ergl looks good to me
